### PR TITLE
Hepmc3tog4, addind the codes that provide a possibility to transfer HepMC3Product from the GEN step to the SIM step 

### PR DESCRIPTION
**Pull request by mkirsano**
This is a continuation of the work described in the issue #36662
The added code allows to transfer the HepMC3 event record to Sim and load particles in Geant4. It is NOT USED if HepMC record is produced in the Genarator part.
Two test configurations are added in Pythia8Interface/test

### CODE REVIEW

**Comment by mkirsano** [comment]
please test

File: IOMC/EventVertexGenerators/src/BaseEvtVtxGenerator.cc [hunk 1]
```diff
@@ -47,20 +49,39 @@ void BaseEvtVtxGenerator::produce(Event& evt, const EventSetup&) {
...
+    //
+    HepMCEvt->applyVtxGen(newVertex(engine));
 
-  // Copy the HepMC::GenEvent
-  HepMC::GenEvent* genevt = new HepMC::GenEvent(*HepUnsmearedMCEvt->GetEvent());
-  std::unique_ptr<edm::HepMCProduct> HepMCEvt(new edm::HepMCProduct(genevt));
-  // generate new vertex & apply the shift
-  //
-  HepMCEvt->applyVtxGen(newVertex(engine));
+    //HepMCEvt->LorentzBoost( 0., 142.e-6 );
```
Comment (civanch) [inline IOMC/EventVertexGenerators/src/BaseEvtVtxGenerator.cc:64]: @mkirsano , this commented line should be removed.

File: IOMC/EventVertexGenerators/src/BaseEvtVtxGenerator.cc [hunk 2]
```diff
@@ -47,20 +49,39 @@ void BaseEvtVtxGenerator::produce(Event& evt, const EventSetup&) {
 
   Handle<HepMCProduct> HepUnsmearedMCEvt;
 
-  evt.getByToken(sourceToken, HepUnsmearedMCEvt);
+  bool found = evt.getByToken(sourceToken, HepUnsmearedMCEvt);
+
+  if (found) {  // HepMC event exists
+
+    // Make a copy
+    HepMC::GenEvent* genevt = new HepMC::GenEvent(*HepUnsmearedMCEvt->GetEvent());
```
Comment (civanch) [inline IOMC/EventVertexGenerators/src/BaseEvtVtxGenerator.cc:57]: @mkirsano , is it really needed making genevt via "new", I would try to create  edm::HepMCProduct directly.

File: IOMC/EventVertexGenerators/src/BaseEvtVtxGenerator.cc [hunk 3]
```diff
@@ -47,20 +49,39 @@ void BaseEvtVtxGenerator::produce(Event& evt, const EventSetup&) {
...
 
-  evt.put(std::move(HepMCEvt));
+  } else {  // no HepMC event, try to get HepMC3 event
+
+    Handle<HepMC3Product> HepUnsmearedMCEvt3;
+    found = evt.getByToken(sourceToken3, HepUnsmearedMCEvt3);
+
+    if (!found)
+      throw cms::Exception("ProductAbsent")
+          << "No HepMCProduct, tried to get HepMC3Product, but it is also absent " << std::endl;
```
Comment (civanch) [inline IOMC/EventVertexGenerators/src/BaseEvtVtxGenerator.cc:77]: std::endl  better to remove.

File: IOMC/EventVertexGenerators/src/BaseEvtVtxGenerator.cc (hunk 1 above)
Comment (mkirsano) [reply IOMC/EventVertexGenerators/src/BaseEvtVtxGenerator.cc]: done

File: IOMC/EventVertexGenerators/src/BaseEvtVtxGenerator.cc (hunk 3 above)
Comment (mkirsano) [reply IOMC/EventVertexGenerators/src/BaseEvtVtxGenerator.cc]: done

**Comment by mkirsano** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by mkirsano** [comment]
+1

File: SimG4Core/Generators/src/Generator3.cc [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+              xi += del * px;
+              yi += del * py;
+            }
+            // check eta cut
+            if ((zi >= Z_lmin) & (zi <= Z_lmax) & (xi * xi + yi * yi < theDecRCut2)) {
+              continue;
+            }
+          }
+          const HepMC3::GenParticle *ppointer = pitr.get();
+          if (fLumiFilter && !fLumiFilter->isGoodForLumiMonitor(ppointer)) {  // MK: this function is always true
```
Comment (tvami) [inline SimG4Core/Generators/src/Generator3.cc:360]: Can you comment on the purpose of this? `isGoodForLumiMonitor` is indeed always true

Comment (mkirsano) [reply SimG4Core/Generators/src/Generator3.cc:360]: This comes from the class Generator, written by somebody else for HepMC, time ago.  May be there is a possibility to provide another, non-trivial function for special cases?

Comment (tvami) [reply SimG4Core/Generators/src/Generator3.cc:360]: Yeah but it's ok to do that in another PR if you prefer

**Comment by tvami** [comment]
+analysis

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**